Create cache directories for cover images only when needed

stweil opened the pull request:
This fixes a potential race condition which could result in a runtime warning ("Warning: mkdir(): File exists") on a new VuFind installation.

The cache directories are only needed when caching is enabled.

It is sufficient to check whether they exist and to create them when a cover image is written to the cache.

stweil committed:
Create cache directories for cover images only when needed

This fixes a potential race condition which could result in a
runtime warning ("Warning: mkdir(): File exists") on a new
VuFind installation.

The cache directories are only needed when caching is enabled.

It is sufficient to check whether they exist and to create them
when a cover image is written to the cache.

Signed-off-by: Stefan Weil <sw@weilnetz.de>

rominail commented:
It currently creates the directory with the file name : `mkdir('/tmp/z_test/sub_test/myfile.txt', 0o755, true);` creates a `myfile.txt` directory

The function would need either to strip the file name before creating the dir or `getCachePath` should do it before concatenating the filename

stweil commented:
> It currently creates the directory with the file name

Oh, sorry, I committed the wrong code. You are right, it must use the directory part of course.

Fixed now.

demiankatz commented on module/VuFind/src/VuFind/Cover/Loader.php:
```diff
@@ -697,6 +689,11 @@ protected function processImageURL($url, $cache = true)
             $finalFile = $cache ? $this->localFile : $tempFile . '.jpg';
 
+            // Make sure that the cache directory exists.
+            if (!is_dir($directory = dirname($tempFile))) {
```
Reviewing this code reveals an existing problem: there is no guarantee that $this->localFile will end with a .jpg extension, so the definition of $tempFile is not guaranteed to create a unique filename. I think we should change the definition above with:

`$tempFile = $this->localFile . uniqid();`

stweil replied:
I wonder why $tempFile is not handled like typical temporary files and written to TEMPDIR instead of the cache directory. Changing this would allow further optimization because the check for directory existence would only be needed when caching is enabled.

demiankatz replied:
I believe the reason for this is that in many systems, the TEMPDIR lives on a separate partition. If we write to TEMPDIR and then move the file to a different location, that may be an expensive copy operation instead of a fast/cheap rename operation. Writing the file to the cache initially avoids that possibility. We could make this a configurable option, though.

stweil replied:
I now rebased the pull request and added a commit which should ensure that $tempFile is unique.

demiankatz commented on module/VuFind/src/VuFind/Cover/Loader.php:
```diff
@@ -697,6 +689,11 @@ protected function processImageURL($url, $cache = true)
 
+            // Make sure that the cache directory exists.
+            if (!is_dir($directory = dirname($tempFile))) {
+                mkdir(dirname($tempFile), 0o755, true);
```
There is an optional `dir_permission` setting in the [Cache] section of config.ini for directories in the cache. I wonder if we should use that here if defined, and default to `0o755` if it is not.

stweil replied:
This means duplicating the code from function `ensureCacheDirectoryExists`. I noticed that this function optionally sets the umask, which has an effect on all newly created directories and files, not only for those in the cache!

stweil replied:
Maybe the `umask` setting in config.ini should be removed. Or at least the comment must be fixed and mention that umask is used for all new files and directories.

demiankatz replied:
Since `ensureCacheDirectoryExists` is a public method, you could add the CacheManager as a dependency of the cover loader, and then you could use it directly without duplicating it.

Good catch regarding the `umask`. This setting was added by the team at University of Chicago many years ago, and I doubt it has been widely used outside of their organization. Maybe @bbusenius or @todolson can comment on the original intent and/or whether it is still in use there. I can think of three possible outcomes, but I'm not sure which is the best since I don't personally use the setting at all:

1.) As you propose, we can keep the existing behavior but change the comment. However, I think if we wanted to do that, we would need to move the umask behavior into the bootstrapper to ensure that a consistent umask is used at all times. I don't really think we want to do this -- but if we did, consistency would be important, rather than having umask behavior act as a side effect of whether or not caching is applied somewhere.

2.) We could narrow down the scope so that umask only applies to directories, by simply storing the original umask before we change it, and then restoring it after the directory is created. This would also require changes to the comment in config.ini, since umask would apply much less broadly.

3.) If we really wanted to apply the umask to all files in the cache, we would likely need to add support methods for writing files into the cache and/or do some customization of the underlying caching framework. This would be a large amount of work, but I think it would be necessary if we wanted the behavior to reflect the documentation as written.

Personally, of these options, I would favor number two as the safest and most practical... but number three is probably the most technically correct. I just don't know if it's worth the effort to implement, especially if nobody is using/asking for it.

bbusenius replied:
I hope @todolson has a better of memory of this than I do. I just looked at the code but don't recall anything about it. All I can say is that we have the `umask = 022` line uncommented in our config.

demiankatz replied:
It does appear that (apart from being refactored to a different place) the current code is essentially unchanged from @todolson's patch committed as https://github.com/vufind-org/vufind/commit/0719861390ea0ff54bb8d352d2beae410c590776.

stweil replied:
So how should we proceed? I suggest a 4th solution: remove all umask related code for release 12 because it was broken right from the beginning, is easy to realize outside of VuFind and is nearly (if at all) unused. Even the "inventors" use `umask = 022` which is not needed because this is the default for root and can be made default for other user accounts, too.

demiankatz replied:
@stweil, that sounds reasonable to me unless @bbusenius or @todolson can think of a reason to object.

We still have a couple of days to get some kind of fix into 11.0 if you think it's urgent, but otherwise, as you say, we can target 12.0 for the next round of breaking changes.

bbusenius replied:
I don't object to this. If there's an unintended consequence, we'll deal with it when it comes up.

demiankatz reviewed (changes requested):
Thanks, @stweil, this refactoring makes sense to me. See below for some suggestions. Let me know if you need help with any implementation details!

stweil committed:
Ensure that temporary cover image file name is unique

The definition of $tempFile was not guaranteed to create a
unique filename because this was only done for files with
a .jpg extension.

Reported-by: Demian Katz <demian.katz@villanova.edu>
Signed-off-by: Stefan Weil <sw@weilnetz.de>

demiankatz reviewed (approved):
Thanks, @stweil! It has taken me a while to review this because every time I look at it, I get distracted by unrelated shortcomings of the cover loader. I think the whole thing would benefit from a redesign in the future. However, I am now convinced that your changes here are at least an incremental improvement.
